### PR TITLE
Changes Web Package Webpack settings

### PR DESCRIPTION
**Pull request by Moejoe90**
**Testing notes**
#p3

### CODE REVIEW

File: packages/web/webpack/webpack.config.prod.js [hunk 1]
```diff
@@ -24,4 +34,101 @@ module.exports = merge.smart(configCommon, {
...
+  output: {
+    filename: "[chunkhash].[name].min.js",
+  },
+  plugins: [
+    new CleanWebpackPlugin(),
+    new MiniCssExtractPlugin({
+      filename: "[contenthash].[name].css",
+    }),
+  ],
+  module: {
```
Comment (arjunu) [inline packages/web/webpack/webpack.config.prod.js]: module.rules can be shared

File: packages/web/webpack/webpack.config.ci.prod.js [hunk 2]
```diff
@@ -1,18 +1,16 @@
...
-const configCommon = require("./webpack.config.common");
+const path = require("path");
 const paths = require("./paths");
 
-// CircleCI is starting to run out of memory due to minimization processes like Terser
-//
+const configCommon = require("./webpack.config.common");
 
+/**
+ * @webpack a special webpack configuration that should run only on CI the minimization process is disabled
```
Comment (desfero) [inline packages/web/webpack/webpack.config.ci.prod.js:11]: can we add some information why do we need to disable the minifier on ci? I still don't think we should do that but don't' have time to research the reason behind build slowdown

File: packages/web/webpack/webpack.config.prod.js [hunk 3]
```diff
@@ -2,10 +2,20 @@ const TerserPlugin = require("terser-webpack-plugin");
...
+const { CleanWebpackPlugin } = require("clean-webpack-plugin");
 
-const configCommon = require("./webpack.config.ci.prod");
+const path = require("path");
+const paths = require("./paths");
 
+const configCommon = require("./webpack.config.common");
+
+/**
+ * @webpack a special webpack configuration that should used on when deploying production code
```
Comment (desfero) [inline packages/web/webpack/webpack.config.prod.js:14]: What's the reason of `@webpack` ? Is it custom modifier for tsdocs?

File: packages/web/webpack/webpack.config.prod.js [hunk 4]
```diff
@@ -2,10 +2,20 @@ const TerserPlugin = require("terser-webpack-plugin");
 const OptimizeCSSAssetsPlugin = require("optimize-css-assets-webpack-plugin");
 const webpack = require("webpack");
 const merge = require("webpack-merge");
+const MiniCssExtractPlugin = require("mini-css-extract-plugin");
+const { CleanWebpackPlugin } = require("clean-webpack-plugin");
 
-const configCommon = require("./webpack.config.ci.prod");
+const path = require("path");
```
Comment (desfero) [inline packages/web/webpack/webpack.config.prod.js:8]: `path` can be moved to thrid-party dependencies section

File: packages/web/webpack/webpack.config.ci.prod.js (hunk 2 above)
Comment (Moejoe90) [reply packages/web/webpack/webpack.config.ci.prod.js:11]: Its not even a build slowdown, its full out of memory issues. 
I can add another line but really there is not much to explain

File: packages/web/webpack/webpack.config.prod.js (hunk 3 above)
Comment (Moejoe90) [reply packages/web/webpack/webpack.config.prod.js:14]: Its a webpack configration why not have it a custom modifier? is there any reason against that

File: packages/web/webpack/webpack.config.common.module.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
+const path = require("path");
+const paths = require("./paths");
+
+export const commonModuleConfig = {
```
Comment (Moejoe90) [inline packages/web/webpack/webpack.config.common.module.js:4]: I didn't want to use `merge.smart` As i'm not sure how it works under the hood. that will require proper research and time

File: packages/design-system/src/components/buttons/Button.tsx [hunk 6]
```diff
@@ -102,7 +102,7 @@ const Button = React.forwardRef<HTMLButtonElement, TButtonProps>(
   ) => {
     const withIconOnly = children === undefined;
 
-    if (process.env.NODE_ENV === "development" || process.env.NF_CYPRESS_RUN === "1") {
```
Comment (desfero) [inline packages/design-system/src/components/buttons/Button.tsx:105]: NF_CYPRESS_RUN is specific env variable for the web and given that it's not useful in the design-system build given it's always is unset

File: packages/web/app/config/getConfig.ts [hunk 7]
```diff
@@ -20,32 +20,38 @@ export interface IConfig {
...
 
-export function getConfig(env: NodeJS.ProcessEnv): IConfig {
-  verifyFeatureFlags(env);
+export function getConfig(): IConfig {
+  verifyFeatureFlags();
 
   return {
     ethereumNetwork: {
-      rpcUrl: getRequiredEnv(env, "NF_RPC_PROVIDER"),
+      rpcUrl: getRequiredEnv(process.env.NF_RPC_PROVIDER, "NF_RPC_PROVIDER"),
```
Comment (desfero) [inline packages/web/app/config/getConfig.ts:28]: `env, "NF_RPC_PROVIDER"` with the following approach webpack is not able to provide a static value for the env variable and it ends up with injecting whole env object as a string

File: packages/design-system/webpack/webpack.config.js [hunk 8]
```diff
@@ -43,7 +44,15 @@ const webpackConfig = (env, argv = {}) => {
     ],
     optimization: {
       minimizer: [
-        new TerserPlugin(),
+        new TerserPlugin({
+          terserOptions: {
+            output: {
+              // do not preserve any kind of comments in the final bundle
```
Comment (arjunu) [inline packages/design-system/webpack/webpack.config.js:50]: so it's on by default?

File: packages/design-system/src/components/buttons/Button.tsx (hunk 6 above)
Comment (arjunu) [reply packages/design-system/src/components/buttons/Button.tsx:105]: we need a way to set `process.env.NF_CYPRESS_RUN` for `test:e2e:dev`

File: packages/web/webpack/webpack.config.prod.js [hunk 9]
```diff
@@ -1,11 +1,20 @@
...
-const configCommon = require("./webpack.config.ci.prod");
+const configCommon = require("./webpack.config.common");
+const paths = require("./paths");
 
+/**
+ * @webpack a special webpack configuration that should used on when deploying production code
+ */
 module.exports = merge.smart(configCommon, {
+  devtool: false,
+  mode: "production",
```
Comment (arjunu) [inline packages/web/webpack/webpack.config.prod.js:17]: this already comes from `NODE_ENV=production webpack`

File: .circleci/config.yml [hunk 10]
```diff
@@ -3,12 +3,12 @@ version: 2.1
 executors:
   docker:
     docker:
-      - image: circleci/node:11.14.0
+      - image: circleci/node:12.16.1
```
Comment (arjunu) [inline .circleci/config.yml:6]: They have 12.16.2 

`https://github.com/CircleCI-Public/circleci-dockerfiles/tree/master/node/images/12.16.2-stretch`

File: packages/web/webpack/webpack.config.common.js [hunk 11]
```diff
@@ -24,9 +24,16 @@ module.exports = {
       template: paths.appHtml,
       favicon: paths.favicon,
     }),
-    new webpack.DefinePlugin({
-      "process.env": applicationEnv,
-    }),
+    // TODO: Find a way to warn if not defined env variable is used in the code
+    new webpack.DefinePlugin(
+      Object.entries(appEnv).reduce(
```
Comment (arjunu) [inline packages/web/webpack/webpack.config.common.js:29]: didn't get what this is for

File: packages/web/webpack/webpack.config.prod.js (hunk 9 above)
Comment (desfero) [reply packages/web/webpack/webpack.config.prod.js:17]: Not exactly true, https://webpack.js.org/configuration/mode/
Please remember that setting NODE_ENV doesn't automatically set mode.
The default is `production` but here I want to emphasize and make it consistent with `webpack.config.dev.js`

File: packages/web/webpack/webpack.config.common.js (hunk 11 above)
Comment (desfero) [reply packages/web/webpack/webpack.config.common.js:33]: done, now there is a comment

File: packages/design-system/webpack/webpack.config.js (hunk 8 above)
Comment (desfero) [reply packages/design-system/webpack/webpack.config.js:50]: nope :(

**Pull request merged**